### ISSUE TITLE
Select and RecvMsg are costly on Windows

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#564]
Client loopback test shows that 32% of CPU time is spent in the "select" (22.5%) and "recvmsg" (9.5%) calls for the Windows client. Not clear whether the same issues also happen in Unix builds.

**Comment by huitema** [comment]
On the server side, "send through server sockets" accounts for 50% of CPU, and "select + receive" for another 10%

**Comment by huitema** [comment]
On linux, there is now an option to do "GSO offload" for UDP sockets. See Fastly's blog: https://www.fastly.com/blog/measuring-quic-vs-tcp-computational-efficiency

**Comment by huitema** [comment]
Fastly's paper cites this paper: https://lwn.net/Articles/188489/

**Comment by huitema** [comment]
The equivalent to the UDP GSO option in windows is UDP_SEND_SEGMENT_SIZE or UDP_RECV_MAX_COALESCED_SIZE. Same behavior as GSO offload. Example of code to test availability from `datapath_winuser.c` in  msquic 

```
#ifdef UDP_SEND_MSG_SIZE
{
    DWORD SegmentSize;
    OptionLength = sizeof(SegmentSize);
    Result =
        getsockopt(
            UdpSocket,
            IPPROTO_UDP,
            UDP_SEND_MSG_SIZE,
            (char*)&SegmentSize,
            &OptionLength);
    if (Result != NO_ERROR) {
        int WsaError = WSAGetLastError();
        QuicTraceLogWarning(
            DatapathQueryUdpSendMsgFailed,
            "[ udp] Query for UDP_SEND_MSG_SIZE failed, 0x%x",
            WsaError);
    } else {
        Datapath->Features |= QUIC_DATAPATH_FEATURE_SEND_SEGMENTATION;
    }
}
#endif

#ifdef UDP_RECV_MAX_COALESCED_SIZE
{
    DWORD UroMaxCoalescedMsgSize = TRUE;
    OptionLength = sizeof(UroMaxCoalescedMsgSize);
    Result =
        getsockopt(
            UdpSocket,
            IPPROTO_UDP,
            UDP_RECV_MAX_COALESCED_SIZE,
            (char*)&UroMaxCoalescedMsgSize,
            &OptionLength);
    if (Result != NO_ERROR) {
        int WsaError = WSAGetLastError();
        QuicTraceLogWarning(
            DatapathQueryRecvMaxCoalescedSizeFailed,
            "[ udp] Query for UDP_RECV_MAX_COALESCED_SIZE failed, 0x%x",
            WsaError);
    } else {
        Datapath->Features |= QUIC_DATAPATH_FEATURE_RECV_COALESCING;
    }
}
#endif
```

**Comment by huitema** [comment]
Implemented the "coalesced" APIs in the windows socket loop. My itself, this appears insufficient:

1) On a loopback test, we achieve 2 to 3 packets per train, despite waking up when 10 packets worth of pacing credits are available. Something else is preventing full trains, probably CWIN.

2) The best performance is observed with Cubic-- 700 Mbps on loopback. BBR achieves 600 to 640 Mbps. This is probably due to poor performance of BBR slow start algorithm on a loopback connection. Debug traces shows the memory increasing exponentially, as if doing slow start, up to about 1/2 of final size. Then, increase looks like "probe_bw" behavior.

3) In theory, even a modest number of packets per train should result in a significant reduction in the cost of the SendMSG API. In practice, this is not happening. Everything behaves as if the API was implemented using a loop on the single message sending code. 

Dang.

**Comment by huitema** [comment]
Also, it appears that the cost is incurred in "send msg start", not in the completion routine. Next step is to consider implementing the "completion port" and "worker threads" architecture for the Windows loop.

**Comment by huitema** [comment]
Why does the train contain so few packets? Because of pacing and congestion control, returning too soon. Measurements show that, if we look at trains stopped after fewer than 4 packets, the two main causes are congestion window reached (less than 1/2 the cases) and pacing bucket empty (almost all other cases). explanations:

* If ACKs are arriving too frequently, CWIN will progress by just a few packets at a time, reducing the size of the trains.
* If the sender is awakened for other reasons than "pacing bucket full", the pacing bucket will only let a small train out.

Next step should be, fix those two issues.

**Comment by huitema** [comment]
Packet train sizes issues were fixed in PR #1102, by using aggressive ACK delays. 

In 64bit release mode, using Cubic, loopback runs at 1.6 Gbps, which is faster than the equivalent run on Linux.
